### PR TITLE
Issue 1343: Shortcode generator support in Classic Editor.

### PR DESCRIPTION
**Pull request by dhaval-parekh**
- laterpay_premium_download
- laterpay_time_passes
- laterpay_subscription_purchase

### CODE REVIEW

File: laterpay/application/Controller/Admin/Post/Metabox.php [hunk 1]
```diff
@@ -33,11 +33,11 @@ public static function get_subscribed_events() {
             ),
             'laterpay_admin_enqueue_styles_post_edit' => array(
                 array( 'laterpay_on_admin_view', 200 ),
-                array( 'load_assets' ),
+                array( 'load_assets', 100 ),
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/Post/Metabox.php]: Did `load_assets` run before `laterpay_on_admin_view`, is that why this change is added?

File: laterpay/application/Controller/Admin/TinyMCE.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
+<?php
+
+/**
+ * Laterpay TinyMCE controller.
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/TinyMCE.php]:
```suggestion
 * LaterPay TinyMCE controller.
```
Could we update the usage to this? we can keep all lowercase in variables and methods.

File: laterpay/application/Controller/Admin/TinyMCE.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    /**
+     * To localize text for shortcode generator.
+     *
+     * @return void
+     */
+    public function localize_script() {
+
+        // Time passes and vouchers data.
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/TinyMCE.php]:
```suggestion
        // Time passes data.
```

File: laterpay/application/Controller/Admin/TinyMCE.php [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+        wp_localize_script(
+            'laterpay-post-edit',
+            'laterpay_shortcode_generator',
+            array(
+                'button'                       => array(
+                    'text' => esc_html__( 'LaterPay ShortCodes', 'laterpay' ),
+                ),
+                'premium_download'             => array(
+                    'title'             => esc_html__( 'LaterPay Premium Download', 'laterpay' ),
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/TinyMCE.php]:
```suggestion
                    'title'             => esc_html__( 'LaterPay Premium Download Box', 'laterpay' ),
```

File: laterpay/application/Core/Bootstrap.php [hunk 5]
```diff
@@ -159,6 +159,9 @@ private function register_admin_actions() {
         $controller = self::get_controller( 'Admin_Advanced' );
         laterpay_event_dispatcher()->add_subscriber( $controller );
 
+        $controller = self::get_controller( 'Admin_TinyMCE' );
+        laterpay_event_dispatcher()->add_subscriber( $controller );
```
Comment (thrijith) [inline laterpay/application/Core/Bootstrap.php]:
```suggestion
        laterpay_event_dispatcher()->add_subscriber( self::get_controller( 'Admin_TinyMCE' ) );
```
Could we remove usage of `$controller` and directly add the controller?

File: laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js [hunk 6]
```diff
@@ -0,0 +1,328 @@
+/**
+ * To add TinyMCE drop down for shortcode generator.
+ *
+ * @package laterpay
+ */
+/*global laterpay_shortcode_generator */
```
Comment (thrijith) [inline laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js]:
```suggestion
/* globals laterpay_shortcode_generator, tinymce, wp */
```
This should be the first line.

File: laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+'use strict';
+
+var laterpay_shortcode_genrator = {
```
Comment (thrijith) [inline laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js]: Could we update the usage to a similar usage in other files?

For e.g: `laterpay/asset_sources/js/laterpay-backend-advanced.js`

Update usage to use iife and pass jQuery as $ so you could define common things earlier and use them later.

File: laterpay/application/Controller/Admin/TinyMCE.php [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+     *
+     * @param LaterPay_Core_Event $event
+     *
+     * @return void
+     */
+    public function add_tinymce_button( LaterPay_Core_Event $event ) {
+
+        $result = $event->get_result();
+
+        $result['laterpay_shortcode_generator'] = sprintf( '%slaterpay-backend-shortcode-generator.js', $this->config->js_url );
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/TinyMCE.php:73]: The script should only load on Post Types allowed for sale, else it would throw an error on load.

`in_array( $post_type, $this->config->get( 'content.enabled_post_types' )`

File: laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+                            editor.windowManager.open( {
+                                title: laterpay_shortcode_generator.time_pass_purchase_button.title,
+                                width: 512,
+                                height: 400,
+                                body: [
+                                    {
+                                        type: 'listbox',
+                                        name: 'id',
+                                        label: laterpay_shortcode_generator.time_pass_purchase_button.id.label,
+                                        values: laterpay_shortcode_generator.time_pass_purchase_button.id.values,
```
Comment (thrijith) [inline laterpay/asset_sources/js/laterpay-backend-shortcode-generator.js]: If there is no time pass / subscription maybe disable the button.

File: laterpay/application/Controller/Admin/TinyMCE.php [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                'value' => $item['id'],
+            );
+        }
+
+        wp_localize_script(
+            'laterpay-post-edit',
+            'laterpay_shortcode_generator',
+            array(
+                'button'                       => array(
+                    'text' => esc_html__( 'LaterPay ShortCodes', 'laterpay' ),
```
Comment (thrijith) [inline laterpay/application/Controller/Admin/TinyMCE.php:124]: Would be possible to use an icon instead?

**Review by thrijith — CHANGES_REQUESTED** [review]

File: laterpay/application/Controller/Admin/Post/Metabox.php (hunk 1 above)
Comment (dhaval-parekh) [reply laterpay/application/Controller/Admin/Post/Metabox.php]: I have change the priority. because it need to run before `LaterPay_Controller_Admin_TinyMCE::localize_script()`
https://github.com/rtCamp/laterpay-wordpress-plugin/pull/23/files#diff-242d3af52dab289cf11c4e6fcfee7c5fR83

So, script will already enqueued before we localize data.

File: laterpay/application/Controller/Admin/TinyMCE.php (hunk 10 above)
Comment (dhaval-parekh) [reply laterpay/application/Controller/Admin/TinyMCE.php:124]: Icon added.
